// rollup-plugin-unassert/dist/rollup-plugin-unassert.es.js
import { createFilter } from 'rollup-pluginutils';
import acorn from 'acorn';
import escodegen from 'escodegen';
import unassertjs from 'unassert';
import convert from 'convert-source-map';
import { transfer } from 'multi-stage-sourcemap';

// Adapted from https://github.com/unassert-js/unassertify/blob/master/index.js
// by Takuto Wada (MIT-licensed)

function handleIncomingSourceMap(originalCode) {
    var commented = convert.fromSource(originalCode);
    if (commented) {
        return commented.toObject();
    }
    return null;
}

function overwritePropertyIfExists(name, from, to) {
    if (from.hasOwnProperty(name)) {
        to.setProperty(name, from[name]);
    }
}
function reconnectSourceMap(inMap, outMap) {
    var mergedRawMap = mergeSourceMap(inMap, outMap.toObject());
    var reMap = convert.fromObject(mergedRawMap);
    overwritePropertyIfExists('sources', inMap, reMap);
    overwritePropertyIfExists('sourceRoot', inMap, reMap);
    overwritePropertyIfExists('sourcesContent', inMap, reMap);
    return reMap;
}

function mergeSourceMap(incomingSourceMap, outgoingSourceMap) {
    if (typeof outgoingSourceMap === 'string' || outgoingSourceMap instanceof String) {
        outgoingSourceMap = JSON.parse(outgoingSourceMap);
    }
    if (!incomingSourceMap) {
        return outgoingSourceMap;
    }
    return JSON.parse(transfer({fromSourceMap: outgoingSourceMap, toSourceMap: incomingSourceMap}));
}

function unassert(options) {
    if ( options === void 0 ) options = {};

    if (options.sourcemap === undefined) {
        options.sourcemap = true;
    }

    var filter = createFilter(
        options.include || ['*.js', '**/*.js'],
        options.exclude
    );

    return {
        transform: function transform(code, id) {
            if (!filter(id)) { return null; }

            return new Promise(function (resolve) {
                var ast = acorn.parse(code, {sourceType: 'module', locations: true});
                var unassertedAst = escodegen.generate(unassertjs(ast), {
                    sourceMap: id,
                    sourceContent: code,
                    sourceMapWithCode: true
                });

                var inMap = options.sourcemap && handleIncomingSourceMap(code);
                var outMap = convert.fromJSON(unassertedAst.map.toString());
                if (inMap) {
                    outMap = reconnectSourceMap(inMap, outMap);
                }

                resolve({
                    code: unassertedAst.code,
                    map: outMap.toObject()
                });
            });
        }
    };
}

export default unassert;
